### PR TITLE
[WFCORE-5032] Elytron subsystem should not depend upon PicketBox

### PR DESCRIPTION
**Pull request by darranl**
Performing some testing provisioning a server without legacy security was causing JACC to fail, this PR is the WildFly Core side of changes to revisit the relationship between Elytron and Legacy Security for JACC handling.  To get clean CI runs this will require the WildFly PR to be merged first:
    https://github.com/wildfly/wildfly/pull/13407

https://issues.redhat.com/browse/WFCORE-5032 - Elytron should not have a dependency on PicketBox.
https://issues.redhat.com/browse/WFCORE-5020 - Ensure the PolicyConfigurationFactory registration happens using service start up and uses the specified module for class loading.
https://issues.redhat.com/browse/WFCORE-5049 - Ensure any Policy is installed before instantiating the PolicyConfigurationFactory.

Starting off with the hold label in place as some additional testing is also being performed but unless issues are found these PRs are ready for review.

### CODE REVIEW

**Comment by darranl** [comment]
Combined Experimental PR https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/212446

**Comment by darranl** [comment]
Experimental build with -Delytron https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/212448

**Comment by darranl** [comment]
Replacement experimental builds after the mixed domain failures:
  https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/213568
  https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/213570

**Comment by darranl** [comment]
The mixed domain test failures on https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/213570 can be ignored, this was a -Delytron run and we do not normally run mixed domain tests on that profile.

File: elytron/src/main/java/org/wildfly/extension/elytron/PolicyDefinitions.java [hunk 1]
```diff
@@ -287,292 +276,131 @@ protected void recordCapabilitiesAndRequirements(OperationContext context, Attri
...
-                }
-                else if (credential instanceof X509CertificateChainPrivateCredential) {
-                    addPrivateCredential(subject, credential.castAs(X509CertificateChainPrivateCredential.class).getCertificateChain());
-                }
-                else {
-                    addPrivateCredential(subject, credential);
-                }
+        try {
+            Thread.currentThread().setContextClassLoader(classLoader);
+            System.setProperty(PolicyConfigurationFactory.class.getName() + ".provider", className);
```
Comment (jmesnil) [inline elytron/src/main/java/org/wildfly/extension/elytron/PolicyDefinitions.java:392]: should this sys prop remain? or should it be discarded at some point?

Comment (darranl) [reply elytron/src/main/java/org/wildfly/extension/elytron/PolicyDefinitions.java:392]: TBH after setting it and triggering the initialisation clearing it has no effect, without restarting the process it is not possible to clear the previous result.

TBH compared to some of the other specs this seems to be a gap in JACC that was can't manually swap in a different implementation / instance.

Comment (jmesnil) [reply elytron/src/main/java/org/wildfly/extension/elytron/PolicyDefinitions.java:392]: ok, thanks for the explanation

**Comment by darranl** [comment]
I can reproduce the failure now, this is the relevant error which is related to the changes:

```
00:09:47,720 ERROR (Controller Boot Thread) [org.jboss.as.controller.management-operation] <AbstractOperationContext.java:1039> WFLYCTL0013: Operation ("add") failed - address: ([
("subsystem" => "elytron"),
("jaspi-configuration" => "test")
]): java.lang.IllegalStateException: Failed to find AuthConfigFactory : org.jboss.security.auth.message.config.JBossAuthConfigFactory
  at javax.security.auth.message.config.AuthConfigFactory.getFactory(AuthConfigFactory.java:227)
  at org.wildfly.security.auth.jaspi.JaspiConfigurationBuilder.register(JaspiConfigurationBuilder.java:106)
  at org.wildfly.extension.elytron.JaspiDefinition$JaspiAddHandler.performRuntime(JaspiDefinition.java:195)
  at org.jboss.as.controller.AbstractAddStepHandler.performRuntime(AbstractAddStepHandler.java:343)
  at org.jboss.as.controller.AbstractAddStepHandler$1.execute(AbstractAddStepHandler.java:164)
  at org.jboss.as.controller.AbstractOperationContext.executeStep(AbstractOperationContext.java:1005)
  at org.jboss.as.controller.AbstractOperationContext.processStages(AbstractOperationContext.java:744)
  at org.jboss.as.controller.AbstractOperationContext.executeOperation(AbstractOperationContext.java:468)
  at org.jboss.as.controller.OperationContextImpl.executeOperation(OperationContextImpl.java:1413)
  at org.jboss.as.controller.ModelControllerImpl.boot(ModelControllerImpl.java:503)
  at org.jboss.as.controller.AbstractControllerService.boot(AbstractControllerService.java:515)
  at org.jboss.as.controller.AbstractControllerService.boot(AbstractControllerService.java:477)
  at org.jboss.as.model.test.ModelTestModelControllerService.boot(ModelTestModelControllerService.java:412)
  at org.jboss.as.controller.AbstractControllerService.boot(AbstractControllerService.java:464)
  at org.jboss.as.controller.AbstractControllerService$1.run(AbstractControllerService.java:416)
  at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassNotFoundException: org.jboss.security.auth.message.config.JBossAuthConfigFactory
  at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
  at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
  at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
  at javax.security.auth.message.config.AuthConfigFactory$LoadAction.run(AuthConfigFactory.java:571)
  at java.base/java.security.AccessController.doPrivileged(Native Method)
  at javax.security.auth.message.config.AuthConfigFactory.getFactory(AuthConfigFactory.java:211)
  ... 15 more
```

I suspect running locally the tests are in a different order so something is initialising the Elytron AuthConfigFactory and leaving it registered, when I run the test in isolation I can reproduce.

**Comment by darranl** [comment]
@jmesnil I have added one more commit to this one, the test failure was genuine - turns out it was hidden by tests running in a different order locally than on CI but the commit should address that issue.

**Comment by darranl** [comment]
The full failures are expected until https://github.com/wildfly/wildfly/pull/13407 is merged to WildFly.

**Comment by jmesnil** [comment]
@darranl thanks

**Pull request merged**